sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -426,7 +426,10 @@ case class In(value: Expression, list: Seq[Expression]) extends Predicate {
  * Optimized version of In clause, when all filter values of In clause are
  * static.
  */
-case class InSet(child: Expression, hset: Set[Any]) extends UnaryExpression with Predicate {
+case class InSet(
+    child: Expression,
+    hset: Set[Any],
+    hsetElemType: DataType) extends UnaryExpression with Predicate {
```

Comment:
Matching internal Catalyst's types to external types is ambiguous. For example,
Long -> Long
Long -> Timestamp

Also type of child can be unknown when InSet has to know Catalyst's type of hset elements.

`hsetElemType` is needed to eliminate the ambiguity

Reply:
Got it.

Reply:
Do you think we can make this `Option[DataType]` because only a few things are ambiguous?

Reply:
We can but if a caller passes None, `InSet` will be not able to infer elem types when `child.dataType` is `NullType` like in this case. `dataType` returns `NullType` if `child` is `PrettyAttribute`.

Reply:
when `hsetElemType` can be different from `child.dataType`?

Reply:
When InSet is created from isInCollection, in that case `child.dataType` is `NullType`. For example, it is NullType in the test https://github.com/apache/spark/pull/28328/files#diff-aa655ba249e00d2591b21cf6a360cf82R886 because child is PrettyAttribute when the `sql` method is called.

Reply:
And `InSet.sql()` is called from `Dataset.select` _.named:
```scala
Project(untypedCols.map(_.named), logicalPlan)
```
The `named` method calls `toPrettySQL(expr)`:
```scala
case expr: Expression => Alias(expr, toPrettySQL(expr))()
```
The `toPrettySQL` method calls `sql`:
```scala
def toPrettySQL(e: Expression): String = usePrettyExpression(e).sql
```